In [1]:
import pandas as pd
import numpy as np

In [2]:
#Importing the labeled dataset available for training the algorithm
# The dataset contains details of the Vehicle insurance holders and response to the cross sales targeting
health_insurance = pd.read_csv('train.csv')

In [3]:
type(health_insurance)

pandas.core.frame.DataFrame

In [4]:
health_insurance.head()

id  Gender  Age  Driving_License  Region_Code  Previously_Insured  \
0   1    Male   44                1         28.0                   0   
1   2    Male   76                1          3.0                   0   
2   3    Male   47                1         28.0                   0   
3   4    Male   21                1         11.0                   1   
4   5  Female   29                1         41.0                   1   

  Vehicle_Age Vehicle_Damage  Annual_Premium  Policy_Sales_Channel  Vintage  \
0   > 2 Years            Yes         40454.0                  26.0      217   
1    1-2 Year             No         33536.0                  26.0      183   
2   > 2 Years            Yes         38294.0                  26.0       27   
3    < 1 Year             No         28619.0                 152.0      203   
4    < 1 Year             No         27496.0                 152.0       39   

   Response  
0         1  
1         0  
2         1  
3         0  
4         0

In [5]:
health_insurance.shape

(381109, 12)

##### There are 381109 records with 12 variables.11 of them can be used for training the data. the Response variable will be used as the outcome

In [6]:
health_insurance.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 381109 entries, 0 to 381108
Data columns (total 12 columns):
 #   Column                Non-Null Count   Dtype  
---  ------                --------------   -----  
 0   id                    381109 non-null  int64  
 1   Gender                381109 non-null  object 
 2   Age                   381109 non-null  int64  
 3   Driving_License       381109 non-null  int64  
 4   Region_Code           381109 non-null  float64
 5   Previously_Insured    381109 non-null  int64  
 6   Vehicle_Age           381109 non-null  object 
 7   Vehicle_Damage        381109 non-null  object 
 8   Annual_Premium        381109 non-null  float64
 9   Policy_Sales_Channel  381109 non-null  float64
 10  Vintage               381109 non-null  int64  
 11  Response              381109 non-null  int64  
dtypes: float64(3), int64(6), object(3)
memory usage: 34.9+ MB


There are 9 columns of numeric datatype and 3 of object type seen. 

Of the numeric datatype, id,Age, Driving_license,Prevously Insured, Vintage and Response are seen to be integers.Region_Code, 
Annual Premium and Policy Sales Channel are seen to be float.

Gender, Vehicle Age and Vehicle Damage are seen to be of object type.

Integers seem to be either IDs or categorical variables. While objects are purely categorical, Floats are purely amounts.


Further analysis to be done to understand these assumptions.

In [7]:
health_insurance.describe()

id            Age  Driving_License    Region_Code  \
count  381109.000000  381109.000000    381109.000000  381109.000000   
mean   190555.000000      38.822584         0.997869      26.388807   
std    110016.836208      15.511611         0.046110      13.229888   
min         1.000000      20.000000         0.000000       0.000000   
25%     95278.000000      25.000000         1.000000      15.000000   
50%    190555.000000      36.000000         1.000000      28.000000   
75%    285832.000000      49.000000         1.000000      35.000000   
max    381109.000000      85.000000         1.000000      52.000000   

       Previously_Insured  Annual_Premium  Policy_Sales_Channel  \
count       381109.000000   381109.000000         381109.000000   
mean             0.458210    30564.389581            112.034295   
std              0.498251    17213.155057             54.203995   
min              0.000000     2630.000000              1.000000   
25%              0.000000    24405.000000             29.000000   
50%              0.000000    31669.000000            133.000000   
75%              1.000000    39400.000000            152.000000   
max              1.000000   540165.000000            163.000000   

             Vintage       Response  
count  381109.000000  381109.000000  
mean      154.347397       0.122563  
std        83.671304       0.327936  
min        10.000000       0.000000  
25%        82.000000       0.000000  
50%       154.000000       0.000000  
75%       227.000000       0.000000  
max       299.000000       1.000000

The ids seem to be unique integer identifiers of the customers. their uniqueness to be determined.

The minimum age of the customer opting for Vehicle insurance is 20 while maximum age is 85. On an average the age of a custoemr opting for vehicle insurance is 38-39.

Driving License as per the assumption is a categorical variable containing only 0s and 1s.99.78% of the customers opting for the insuurance have a driving license.

Region_code as per assumption is also a categorical variable representing the region from which the customer is opting for the insurance. there seem to be 53 regions with codes ranging from 0 to 52.

Previously insured is also a categorical variable with 1 stnding for 'Yes' and 0 standing fr 'No'. 45.82% of the customers are seen to hold an insurance previously.

Annual Premium represents the amount to be paid annually for the insurance. it is seen to range from 2630 to as high as 540,165 dollars. On an average a customer pays a premium of 30,564 for annual premium

Policy sales channel is a categorical variable ranging from 1 to 163. There seem to be 163 channels of reaching out to customers.

Vintage representing the number of days a customer has been with the company. the most recent customer has been with the company for 10 days and the oldest one for 299 days. on a average a customer has been for 154 days.

The response can be either 1 for 'Yes' or a 0 for 'No'. 12% of the vehicle insurance holders seem to have responded positively at opting for a health insurance

In [9]:
health_insurance['id'].nunique()

381109

The ids are unique for all the rows. Hence the data is unique at customer id level.

It is time to check for the missing/null values and values that do not fit in

In [10]:
nullcounts = health_insurance.isnull().sum().sort_values(ascending=False)

In [11]:
nullcounts

Response                0
Vintage                 0
Policy_Sales_Channel    0
Annual_Premium          0
Vehicle_Damage          0
Vehicle_Age             0
Previously_Insured      0
Region_Code             0
Driving_License         0
Age                     0
Gender                  0
id                      0
dtype: int64

There are no missing values seen for all the variables considered!

We have to check for any values that do not make sense.

In [12]:
health_insurance['Gender'].value_counts()

Male      206089
Female    175020
Name: Gender, dtype: int64

The data is neatly categorized as Males and Females. However it is of Integer format. It needs to be formatted as string and the values can be shortened.

In [13]:
health_insurance['Gender'] = health_insurance['Gender'].astype(str)
health_insurance['Gender'].replace({'Male':'M', 'Female': 'F'}, inplace =True)

In [14]:
health_insurance['Gender'].unique()

array(['M', 'F'], dtype=object)

In [15]:
health_insurance['Age'].value_counts()

24    25960
23    24256
22    20964
25    20636
21    16457
      ...  
81       56
82       29
83       22
84       11
85       11
Name: Age, Length: 66, dtype: int64

No Outliers or missing values seen in Age. The range of age is also practically placed from 24 to 85.

In [16]:
health_insurance['Driving_License'].value_counts()

1    380297
0       812
Name: Driving_License, dtype: int64

The customers either have a driving license or not have one. The values 0 and 1 make sense.

In [17]:
health_insurance['Previously_Insured'].value_counts()

0    206481
1    174628
Name: Previously_Insured, dtype: int64

The Boolean values aptly describe if a customer has bought an insurance previously or not.

In [18]:
health_insurance['Region_Code'].value_counts()

28.0    106415
8.0      33877
46.0     19749
41.0     18263
15.0     13308
30.0     12191
29.0     11065
50.0     10243
3.0       9251
11.0      9232
36.0      8797
33.0      7654
47.0      7436
35.0      6942
6.0       6280
45.0      5605
37.0      5501
18.0      5153
48.0      4681
14.0      4678
39.0      4644
10.0      4374
21.0      4266
2.0       4038
13.0      4036
7.0       3279
12.0      3198
9.0       3101
27.0      2823
32.0      2787
43.0      2639
17.0      2617
26.0      2587
25.0      2503
24.0      2415
38.0      2026
0.0       2021
16.0      2007
31.0      1960
23.0      1960
20.0      1935
49.0      1832
4.0       1801
34.0      1664
19.0      1535
22.0      1309
40.0      1295
5.0       1279
1.0       1008
44.0       808
42.0       591
52.0       267
51.0       183
Name: Region_Code, dtype: int64

There are 53 region codes seen. However the codes would better depict their regions if the decimals are removed.

In [19]:
pd.set_option('precision', 0)

In [20]:
health_insurance['Region_Code'].value_counts()

28    106415
8      33877
46     19749
41     18263
15     13308
30     12191
29     11065
50     10243
3       9251
11      9232
36      8797
33      7654
47      7436
35      6942
6       6280
45      5605
37      5501
18      5153
48      4681
14      4678
39      4644
10      4374
21      4266
2       4038
13      4036
7       3279
12      3198
9       3101
27      2823
32      2787
43      2639
17      2617
26      2587
25      2503
24      2415
38      2026
0       2021
16      2007
31      1960
23      1960
20      1935
49      1832
4       1801
34      1664
19      1535
22      1309
40      1295
5       1279
1       1008
44       808
42       591
52       267
51       183
Name: Region_Code, dtype: int64

The values are now devoid of any decimal points

In [21]:
health_insurance['Annual_Premium'].value_counts()

2630     64877
69856      140
39008       41
38287       38
45179       38
         ...  
62326        1
59733        1
55934        1
75387        1
53346        1
Name: Annual_Premium, Length: 48838, dtype: int64

There are no outliers in annual premiums and the values seem to fall in a practical range

In [22]:
health_insurance['Policy_Sales_Channel'].value_counts()

152    134784
26      79700
124     73995
160     21779
156     10661
        ...  
144         1
149         1
84          1
143         1
43          1
Name: Policy_Sales_Channel, Length: 155, dtype: int64

The channels have been masked using integer codes and seem to be following a pattern

In [23]:
health_insurance['Vintage'].value_counts()

256    1418
73     1410
282    1397
158    1394
187    1392
       ... 
205    1235
89     1234
32     1230
224    1227
277    1226
Name: Vintage, Length: 290, dtype: int64

The number of days for which insurance has been held seems to range from 10 to 299.

In [24]:
health_insurance['Vehicle_Age'].value_counts()

1-2 Year     200316
< 1 Year     164786
> 2 Years     16007
Name: Vehicle_Age, dtype: int64

The Vehicle Age has three categories and better be formatted as object instead of integer.

In [25]:
health_insurance['Vehicle_Age'] = health_insurance['Vehicle_Age'].astype(str)

In [26]:
health_insurance['Vehicle_Age'].unique()

array(['> 2 Years', '1-2 Year', '< 1 Year'], dtype=object)

The Age buckets are now formatted as strings

In [27]:
health_insurance['Vehicle_Damage'].value_counts()

Yes    192413
No     188696
Name: Vehicle_Damage, dtype: int64

The data type is seen to be of integer while the values are 'Yes' and 'No.'
It will be convinient to have the Boolean values as 0 and 1.

In [28]:
health_insurance['Vehicle_Damage'].replace({'Yes':1, 'No':0}, inplace=True)

In [29]:
health_insurance['Vehicle_Damage'].value_counts()

1    192413
0    188696
Name: Vehicle_Damage, dtype: int64

In [30]:
health_insurance['Response'].value_counts()

0    334399
1     46710
Name: Response, dtype: int64

The responses are Yes and No represented by 1 and 0 respectively. It is also of integer format and hence no further changes required.

Dataframe can now be exported for usage in further analysis.

In [31]:
health_insurance.to_csv('Cleaned_Insurance_data.csv')